src/friendly_errors/engine.js

```diff
@@ -0,0 +1,569 @@
+          const match = issue.message.match(/Input not instance of (\w+)/);
+          if (match) expectedTypes.add(match[1]);
+          actualType = undefined;
+        }
+      }
+    });
+
+    if (expectedTypes.size > 0) {
+      if (error.path?.length > 0 && args[error.path[0]] instanceof Promise)  {
+        message += 'Did you mean to put `await` before a loading function? ' +
+          'An unexpected Promise was found. ';
+        isVersionError = true;
+      }
+
+      const expectedTypesStr = Array.from(expectedTypes).join(' or ');
+      const position = error.path.join('.');
+
+      message += buildTypeMismatchMessage(
+        actualType, expectedTypesStr, position
+      );
+    }
+
+    return message;
+  };
+
+  switch (currentError.code) {
+    case 'invalid_union': {
+      processUnionError(currentError);
+      break;
+    }
+    case 'too_small': {
+      const minArgs = currentError.minimum;
+      message += `Expected at least ${minArgs} argument${minArgs > 1 ? 's' : ''}, but received fewer`;
+      break;
+    }
+    case 'invalid_type': {
+      message += buildTypeMismatchMessage(currentError.message.split(', received ')[1], currentError.expected, currentError.path.join('.'));
+      break;
+    }
+    case 'too_big': {
+      const maxArgs = currentError.maximum;
+      message += `Expected at most ${maxArgs} argument${maxArgs > 1 ? 's' : ''}, but received more`;
+      break;
+    }
+    default: {
+      console.log('Zod error object', currentError);
+    }
+  }
+
+  // Let the user know which function is generating the error.
+  message += ` in ${func}().`;
+
+  // Generates a link to the documentation based on the given function name.
+  // TODO: Check if the link is reachable before appending it to the error
+  // message.
+  const generateDocumentationLink = func => {
+    const { funcName, funcClass } = extractFuncNameAndClass(func);
+    const p5BaseUrl = 'https://p5js.org/reference';
+    const url = `${p5BaseUrl}/${funcClass}/${funcName}`;
+
+    return url;
+  };
+
+  if (currentError.code === 'too_big' || currentError.code === 'too_small') {
+    const documentationLink = generateDocumentationLink(func);
+    message += ` For more information, see ${documentationLink}.`;
+  }
+
+  if (isVersionError) {
+    p5._error(this, message);
+  } else {
+    console.log(message);
+  }
+  return message;
+};
+
+/**
+ * Runs parameter validation by matching the input parameters to Zod schemas
+ * generated from the parameter data from `docs/parameterData.json`.
+ *
+ * @private
+ * @param {String} func - Name of the function.
+ * @param {Array} args - User input arguments.
+ * @returns {Object} The validation result.
+ * @returns {Boolean} result.success - Whether the validation was successful.
+ * @returns {any} [result.data] - The parsed data if validation was successful.
+ * @returns {String} [result.error] - The validation error message if validation has failed.
+ */
+function validate(func, args) {
```

Comment:
Do we use the versions of the functions in this file or do we still use the param_validator.js versions? Might want to remove one of them